src/backend/distributed/executor/tuple_destination.c

```diff
@@ -86,7 +93,28 @@ TupleStoreTupleDestPutTuple(TupleDestination *self, Task *task,
 							HeapTuple heapTuple, uint64 tupleLibpqSize)
 {
 	TupleStoreTupleDestination *tupleDest = (TupleStoreTupleDestination *) self;
+
+	/*
+	 * Remote execution sets tupleLibpqSize, however it is 0 for local execution. We prefer
+	 * to use tupleLibpqSize for  the remote execution because that reflects the exact data
+	 * transfer size over the network. For local execution, we rely on the size of the
+	 * tuple.
+	 */
+	uint64 dataSize =
+		tupleLibpqSize > 0 ? tupleLibpqSize : HeapTupleHeaderGetDatumLength(heapTuple);
+	TupleDestinationStats *tupleDestinationStats = self->tupleDestinationStats;
+	if (SubPlanLevel > 0 && tupleDestinationStats != NULL)
+	{
+		tupleDestinationStats->totalIntermediateResultSize += dataSize;
+		if (CheckIfSizeLimitIsExceeded(tupleDestinationStats))
```

Comment:
We can possibly have EnsureIntermediateSizeLimitNotExceeded that both does the check and error if the check is not satisfied. Something like:

```c
static void EnsureIntermediateSizeLimitNotExceeded(DistributedExecutionStats *executionStats) {
    if (!executionStats) {
         return;
    }
	if (!SubPlanLevel || MaxIntermediateResult < 0)
	{
		return;
	}

	uint64 maxIntermediateResultInBytes = MaxIntermediateResult * 1024L;
	if (executionStats->totalIntermediateResultSize < maxIntermediateResultInBytes)
	{
		return;
	}

		ereport(ERROR, (errmsg("the intermediate result size exceeds "
						   "citus.max_intermediate_result_size (currently %d kB)",
						   MaxIntermediateResult),
					errdetail("Citus restricts the size of intermediate "
							  "results of complex subqueries and CTEs to "
							  "avoid accidentally pulling large result sets "
							  "into once place."),
					errhint("To run the current query, set "
							"citus.max_intermediate_result_size to a higher"
							" value or -1 to disable.")));
}

```
That would possibly simplify the code.